File: frontend/src/lib/maintenance.ts
/**
 * Maintenance 前端展示工具（Sprint 5，展示层，非后端状态机事实源）：
 * - 工单状态 / 分类 / 严重度 / 来源展示元数据（标签 + 徽标颜色）
 * - Active Blocking 状态集合（与后端 BLOCKING_STATUSES 一致，仅用于展示）
 */

import type {
  MaintenanceCategory,
  MaintenanceSeverity,
  MaintenanceSource,
  MaintenanceWorkOrderStatus,
} from "@/lib/api/types";
import type { StatusMeta } from "@/lib/status";

export const MWO_STATUS_META: Record<MaintenanceWorkOrderStatus, StatusMeta> = {
  OPEN: { label: "待处理", badge: "bg-amber-100 text-amber-800 ring-amber-300" },
  ASSIGNED: {
    label: "已派工",
    badge: "bg-sky-100 text-sky-800 ring-sky-300",
  },
  IN_PROGRESS: {
    label: "维修中",
    badge: "bg-blue-100 text-blue-800 ring-blue-300",
  },
  RESOLVED: {
    label: "待验收",
    badge: "bg-violet-100 text-violet-800 ring-violet-300",
  },
  COMPLETED: {
    label: "已完成",
    badge: "bg-emerald-100 text-emerald-800 ring-emerald-300",
  },
  CANCELLED: {
    label: "已取消",
    badge: "bg-slate-200 text-slate-700 ring-slate-400",
  },
};

export const MWO_CATEGORY_LABELS: Record<MaintenanceCategory, string> = {
  ELECTRICAL: "电气",
  PLUMBING: "给排水",
  HVAC: "空调",
  LOCK: "门锁",
  BATHROOM: "卫生间",
  FURNITURE: "家具",
  APPLIANCE: "电器",
  NETWORK: "网络",
  FINISHING: "装修",
  OTHER: "其它",
};

export const MWO_SEVERITY_META: Record<MaintenanceSeverity, StatusMeta> = {
  LOW: { label: "低", badge: "bg-slate-100 text-slate-700 ring-slate-300" },
  MEDIUM: { label: "中", badge: "bg-amber-100 text-amber-800 ring-amber-300" },
  HIGH: { label: "高", badge: "bg-orange-100 text-orange-800 ring-orange-300" },
  CRITICAL: { label: "紧急", badge: "bg-red-100 text-red-700 ring-red-300" },
};

export const MWO_SOURCE_LABELS: Record<MaintenanceSource, string> = {
  MANUAL: "手动报修",
  FRONT_DESK: "前台报修",
  HOUSEKEEPING: "保洁报修",
  PRE_OPENING: "开业检查",
};

export const MWO_STATUSES = Object.keys(
  MWO_STATUS_META,
) as MaintenanceWorkOrderStatus[];

/** Active Blocking 状态（Sprint 5 §11，与后端 BLOCKING_STATUSES 一致；
 *  RESOLVED 仍阻断 —— 维修完成 ≠ 酒店验收通过） */
export const MWO_BLOCKING_STATUSES: MaintenanceWorkOrderStatus[] = [
  "OPEN",
  "ASSIGNED",
  "IN_PROGRESS",
  "RESOLVED",
];

/** 工单是否仍处于阻断状态（仅展示层判断） */
export function isActiveBlocking(
  status: MaintenanceWorkOrderStatus,
  blocksRoom: boolean,
): boolean {
  return blocksRoom && MWO_BLOCKING_STATUSES.includes(status);
}

File: backend/app/models/maintenance.py
"""MaintenanceWorkOrder（维修工单）模型（Sprint 5）。

Maintenance Work Order = 独立业务领域：Issue Reported → Work Order → Assign →
Repair → Resolve → Verify / Rework → Complete → Room Ready。

领域原则（Sprint 5 §2）：
    Maintenance Status ≠ Room Occupancy Status ≠ Cleaning Status
维修工单作为第三个独立业务领域，绝不把维修状态塞进 Room 单状态。

关键语义（Sprint 5 §11/§21/§22/§23）：
- blocks_room = true 表示该 Active 工单阻止客房接受新的住宿业务；
  Active Blocking 状态 = OPEN / ASSIGNED / IN_PROGRESS / RESOLVED
  （RESOLVED 仍阻断：维修完成 ≠ 酒店验收通过）。
- 同一 Room 允许多张 Active 工单（不做 active-per-room 唯一约束）。
- 解除 Maintenance 造成的 OOS 必须同时满足：
    active blocking MWO count == 0
    AND Room.occupancy_status == out_of_service
    AND Room.unavailability_source == MAINTENANCE

PII（Sprint 5 §30）：不关联 Guest / Reservation / Stay，不复制 Guest PII。
"""

import enum
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base
from app.models.room import Room
from app.models.user import User


class MaintenanceWorkOrderStatus(str, enum.Enum):
    OPEN = "OPEN"
    ASSIGNED = "ASSIGNED"
    IN_PROGRESS = "IN_PROGRESS"
    RESOLVED = "RESOLVED"
    COMPLETED = "COMPLETED"
    CANCELLED = "CANCELLED"


class MaintenanceCategory(str, enum.Enum):
    ELECTRICAL = "ELECTRICAL"
    PLUMBING = "PLUMBING"
    HVAC = "HVAC"
    LOCK = "LOCK"
    BATHROOM = "BATHROOM"
    FURNITURE = "FURNITURE"
    APPLIANCE = "APPLIANCE"
    NETWORK = "NETWORK"
    FINISHING = "FINISHING"
    OTHER = "OTHER"


class MaintenanceSeverity(str, enum.Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class MaintenanceSource(str, enum.Enum):
    MANUAL = "MANUAL"
    FRONT_DESK = "FRONT_DESK"
    HOUSEKEEPING = "HOUSEKEEPING"
    PRE_OPENING = "PRE_OPENING"


class MaintenanceWorkOrder(Base):
    __tablename__ = "maintenance_work_orders"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    work_order_no: Mapped[str] = mapped_column(
        String(32), unique=True, nullable=False
    )
    room_id: Mapped[int] = mapped_column(
        ForeignKey("rooms.id", ondelete="RESTRICT"), nullable=False, index=True
    )
    category: Mapped[MaintenanceCategory] = mapped_column(
        Enum(
            MaintenanceCategory,
            name="mwo_category",
            values_callable=lambda members: [m.value for m in members],
        ),
        nullable=False,
        index=True,
    )
    severity: Mapped[MaintenanceSeverity] = mapped_column(
        Enum(
            MaintenanceSeverity,
            name="mwo_severity",
            values_callable=lambda members: [m.value for m in members],
        ),
        nullable=False,
        default=MaintenanceSeverity.MEDIUM,
        server_default=MaintenanceSeverity.MEDIUM.value,
        index=True,
    )
    status: Mapped[MaintenanceWorkOrderStatus] = mapped_column(
        Enum(
            MaintenanceWorkOrderStatus,
            name="mwo_status",
            values_callable=lambda members: [m.value for m in members],
        ),
        nullable=False,
        default=MaintenanceWorkOrderStatus.OPEN,
        server_default=MaintenanceWorkOrderStatus.OPEN.value,
        index=True,
    )
    source: Mapped[MaintenanceSource] = mapped_column(
        Enum(
            MaintenanceSource,
            name="mwo_source",
            values_callable=lambda members: [m.value for m in members],
        ),
        nullable=False,
        default=MaintenanceSource.MANUAL,
        server_default=MaintenanceSource.MANUAL.value,
        index=True,
    )
    blocks_room: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(String(2000))
    reported_by_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    assigned_to_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True
    )
    verified_by_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    resolution_notes: Mapped[str | None] = mapped_column(String(2000))
    verification_notes: Mapped[str | None] = mapped_column(String(2000))
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    verified_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    cancelled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    updated_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    room: Mapped[Room] = relationship()
    assigned_to: Mapped[User | None] = relationship(
        foreign_keys=[assigned_to_user_id]
    )
    reported_by: Mapped[User | None] = relationship(
        foreign_keys=[reported_by_user_id]
    )
    verified_by: Mapped[User | None] = relationship(
        foreign_keys=[verified_by_user_id]
    )
